x-pack/plugins/uptime/public/state/reducers/synthetics.ts

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { createAction, handleActions, Action } from 'redux-actions';
+import {
+  isScreenshotBlockDoc,
+  ScreenshotBlockDoc,
+} from '../../../common/runtime_types/ping/synthetics';
+
+export interface PendingBlock {
+  status: 'pending' | 'loading';
+}
+
+export function isPendingBlock(data: unknown): data is PendingBlock {
+  return ['pending', 'loading'].some((s) => s === (data as PendingBlock)?.status);
+}
+export type StoreScreenshotBlock = ScreenshotBlockDoc | PendingBlock;
+export interface ScreenshotBlockCache {
+  [hash: string]: StoreScreenshotBlock;
+}
+
+export interface CacheHitCount {
+  hash: string;
+  hitTime: number;
+}
+
+export interface SyntheticsReducerState {
+  blocks: ScreenshotBlockCache;
+  cacheSize: number;
+  hitCount: CacheHitCount[];
+}
+
+export interface PutBlocksPayload {
+  blocks: ScreenshotBlockDoc[];
+}
+
+// this action denotes a set of blocks is required
+export const fetchBlocksAction = createAction<string[]>('FETCH_BLOCKS');
+// this action denotes a request for a set of blocks is in flight
+export const setBlockLoadingAction = createAction<string[]>('IN_FLIGHT_BLOCKS_ACTION');
+// block data has been received, and should be added to the store
+export const putBlocksAction = createAction<PutBlocksPayload>('PUT_SCREENSHOT_BLOCKS');
+// updates the total size of the image blob data cached in the store
+export const putCacheSize = createAction<number>('PUT_CACHE_SIZE');
+// keeps track of the most-requested blocks
+export const updateHitCountsAction = createAction<string[]>('UPDATE_HIT_COUNTS');
+// reduce the cache size to the value in the action payload
+export const pruneCacheAction = createAction<number>('PRUNE_SCREENSHOT_BLOCK_CACHE');
+
+const initialState: SyntheticsReducerState = {
+  blocks: {},
+  cacheSize: 0,
+  hitCount: [],
+};
+
+// using `any` here because `handleActions` is not set up well to handle the multi-type
+// nature of all the actions it supports. redux-actions is looking for new maintainers https://github.com/redux-utilities/redux-actions#looking-for-maintainers
+// and seems that we should move to something else like Redux Toolkit.
+export const syntheticsReducer = handleActions<
+  SyntheticsReducerState,
+  string[] & PutBlocksPayload & number
+>(
+  {
+    /**
+     * When removing blocks from the cache, we receive an action with a number.
+     * The number equates to the desired ceiling size of the cache. We then discard
+     * blocks, ordered by the least-requested. We continue dropping blocks until
+     * the newly-pruned size will be less than the ceiling supplied by the action.
+     */
+    [String(pruneCacheAction)]: (state, action: Action<number>) => {
```

Comment:
I'll do this extract the logic into separate functions that are called by the reducer, since I agree we should try to keep the actual switch as clean as we can. I don't think dispatching a separate action is essential in this case. LMK if there's any compelling reason we should. Will link the commit here when I'm done.

Reply:
https://github.com/elastic/kibana/pull/103979/commits/e4fbf2e4b3944f6dad2b90b07a5b467bf704c87c